### PR TITLE
Base64 encode label values in the index.

### LINKED ISSUE #210: Null byte in labels and metric names should be acceptable
They are a valid utf-8 character.

We should validate data based on: https://prometheus.io/docs/concepts/data_model/

### PR DESCRIPTION
**Pull request by tomwilkie**
Fixes #210

Encode label values in base64 in the index.  Query for both base64 encoded and non-base64 encoded values upto a certain date, after which only query to base64 encoded values.

### CODE REVIEW

**Comment by jml** [comment]
```
# github.com/weaveworks/cortex/util

util/validate.go:6: imported and not used: "strings"
```

**Comment by tomwilkie** [comment]
Just fixed that, thanks.

File: chunk/chunk_store.go [hunk 1]
```diff
@@ -190,15 +197,125 @@ func hashValue(userID, bucket string, metricName model.LabelValue) string {
 	return fmt.Sprintf("%s:%s:%s", userID, bucket, metricName)
 }
 
-func rangeValue(label model.LabelName, value model.LabelValue, chunkID string) ([]byte, error) {
-	return lex.Encode(string(label), string(value), chunkID)
+func rangeValue(label model.LabelName, value model.LabelValue, chunkID string) []byte {
+	// encoded length is len(label) + 1 + len(encoded(value)) + 1 + len(chunkID) + 1 + version byte + 1
```
Comment (jml) [inline chunk/chunk_store.go]: Took me a while to parse this comment. I think it'd be more helpful to put a comment above the `length =` line explaining where `5` comes from. e.g.

```golang
// 5 = 4 field terminators + 1 version bytes
```

Or whatever.

File: chunk/chunk_store.go [hunk 2]
```diff
@@ -190,15 +197,125 @@ func hashValue(userID, bucket string, metricName model.LabelValue) string {
...
+		output = make([]byte, length, length)
+		i      = 0
+	)
+	next := func(n int) []byte {
+		result := output[i : i+n]
+		i += n + 1
+		return result
+	}
+	copy(next(labelLen), labelBytes)
+	base64.RawStdEncoding.Encode(next(valueLen), valueBytes)
```
Comment (jml) [inline chunk/chunk_store.go]: We're encoding this twice. Is this a problem?

File: chunk/chunk_store.go [hunk 3]
```diff
@@ -491,59 +605,78 @@ func (c *AWSStore) lookupChunksForMetricName(ctx context.Context, userID string,
...
+			log.Errorf("Error querying DynamoDB: %v", err)
+			return nil, err
+		} else if processingError != nil {
+			log.Errorf("Error processing DynamoDB response: %v", processingError)
+			return nil, processingError
+		}
+
+		sort.Sort(ByID(chunkSet))
+		return chunkSet, nil
+	}
```
Comment (jml) [inline chunk/chunk_store.go]: I'd strongly consider making this a separate method/function, and just pass through all the values it's closing over.

File: chunk/chunk_store.go (hunk 2 above)
Comment (jml) [inline chunk/chunk_store.go]: You explained in person. We're not.

**Comment by tomwilkie** [comment]
> Can we add a test for a query that crosses the flag date?

This PR deletes the code to write the old values, so I'd have to resurrect that... and I don't wanna...

**Comment by tomwilkie** [comment]
@jml verbally LGTMd

**Pull request merged**